### PR TITLE
newlogs endpoint

### PR DESCRIPTION
**Pull request by giggsoff**
Newlogs endpoint implementation for https://github.com/lf-edge/eve/pull/1667

We need to discuss https://github.com/lf-edge/eve/pull/1667#discussion_r531123817

The main question for me here, we can completely abandon LogBundle inside Adam and remove previous implementation of /logs. Or we can keep LogBundle like I do here and feed it with some portions of data (MaxLogEntryInBundle).

Signed-off-by: Petr Fedchenkov <giggsoff@gmail.com>

### CODE REVIEW

File: pkg/server/apiHandler.go [hunk 1]
```diff
@@ -301,6 +304,65 @@ func (h *apiHandler) logs(w http.ResponseWriter, r *http.Request) {
...
+	}
+	h.recordClient(u, r)
+	gr, err := gzip.NewReader(r.Body)
+	if err != nil {
+		log.Printf("error gzip.NewReader: %v", err)
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+	logBundleFromHeader := &logs.LogBundle{}
+	if err := json.Unmarshal([]byte(gr.Comment), logBundleFromHeader); err != nil {
```
Comment (deitch) [inline pkg/server/apiHandler.go:323]: I am pretty sure that [gzip.Reader](https://golang.org/pkg/compress/gzip/#Reader) has a property [Header](https://golang.org/pkg/compress/gzip/#Header) which in turn has a property `Comment`, so `gr.Comment` should be incorrect, but `gr.Header.Comment` should work.

File: pkg/driver/common/common.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ import (
 const (
 	KB = 1024
 	MB = 1024 * KB
+	MaxLogEntryInBundle = 10
```
Comment (deitch) [inline pkg/driver/common/common.go:27]: Where is this from? How do you know that the max in a `LogBundle` is 10?

Comment (deitch) [reply pkg/driver/common/common.go:27]: Also, I don't think this goes in `pkg/driver/common`, as it is not common to the driver (between adam server and backend) , but to the API protocol (between adam server and device).

Comment (giggsoff) [reply pkg/driver/common/common.go:27]: This is just my suggestion: we can have hundreds of LogEntry inside LogBundle (gzip). We can split them into several messages inside redis (with at most MaxLogEntryInBundle).

Comment (deitch) [reply pkg/driver/common/common.go:27]: Oh so this is the max that will get written to the backend, not the max that will be received?

Why would we do it any differently than sent by the device? If it sends us 50 in a bundle, send 50 to the backend?

File: pkg/server/apiHandler.go (hunk 1 above)
Comment (giggsoff) [reply pkg/server/apiHandler.go:323]: Hmm, I thing, they have the same value. You can see the example on top of godoc [gzip](https://golang.org/pkg/compress/gzip) (inside the Overview section).

Comment (deitch) [reply pkg/server/apiHandler.go:323]: Ah, you are right. It is an embedded struct. I missed that. Never mind then, all good. 🤦🏻‍♂️

File: pkg/driver/common/common.go (hunk 2 above)
Comment (giggsoff) [reply pkg/driver/common/common.go:27]: Device send us bundle with tons of LogEntry on the principle of accumulation of gzip (logrotate with size/time trigger). But we can split them by our own logic inside controller (10 to the first LogBundle, 10 to the second...). Again, this is my suggestion, I can remove it if it adds confusion.

Comment (deitch) [reply pkg/driver/common/common.go:27]: So I think you are saying that, in the past, it sent a LogBundle with a reasonable range of LogEntry. We don't really care about the LogBundle, just about the many LogEntry.

Now that it gzips, it might send a lot more at once, and we don't want to store them as they come. 

Now that I understand what you are doing better (thanks for the patience explaining it), here is my thought. 

We don't actually use the LogBundle when reading, do we? We just pass it on as a structure to the underlying device driver, and pull it out when someone asks.

When someone reads logs, all they really care about is the LogEntry. If there are 500 LogEntry, they don't care if it came as one LogBundle with 500 LogEntry, or 10 LogBundle with 50 LogEntry each, or 500 LogBundle with 1 LogEntry each. They just want to see the 500 lines of logs.

I think we shouldn't bother splitting it up any more. We can gunzip them, but then just put them in as-is. If the device sent us 500 in a single LogBundle, then we store it that way.

In the (very near) future, we may want to take a good look and say, "no one cares about LogBundle when reading from Adam; they just care about the 500 lines, so let's store LogEntry and ignore LogBundle entirely." There may be impact to doing that - metadata and such that we care about - but for now, I think I would avoid trying to be "smarter" than the controller's bundling.

File: pkg/server/apiHandler.go [hunk 3]
```diff
@@ -301,6 +304,65 @@ func (h *apiHandler) logs(w http.ResponseWriter, r *http.Request) {
...
+func (h *apiHandler) newLogs(w http.ResponseWriter, r *http.Request) {
+	cert := getClientCert(r)
+	u, err := h.manager.DeviceCheckCert(cert)
+	if err != nil {
+		log.Printf("error checking device cert: %v", err)
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+	h.recordClient(u, r)
+	gr, err := gzip.NewReader(r.Body)
```
Comment (naiming-zededa) [inline pkg/server/apiHandler.go:316]: need to add something like, since the newlog message is only sent in APIV2:
	sm := &zauth.AuthContainer{}
	err := proto.Unmarshal(c, sm)
	data := sm.ProtectedPayload.GetPayload()

Comment (giggsoff) [reply pkg/server/apiHandler.go:316]: It works fine without any modifications for APIV1, and I checked it. Actually, AuthContainer logic depends on variable [inside ZedCloudContext](https://github.com/naiming-zededa/eve/blob/f5db03d68191618031c3b3b993f5ba8784667a36/pkg/pillar/zedcloud/send.go#L399). You build the correct url for device logs [here](https://github.com/naiming-zededa/eve/blob/f5db03d68191618031c3b3b993f5ba8784667a36/pkg/pillar/cmd/loguploader/loguploader.go#L355). 
But I cannot figure out what is stopping you from creating the correct link for application logs (pass `zedcloudCtx.V2API` instead of `true` [here](https://github.com/naiming-zededa/eve/blob/naiming-newlog-infra/pkg/pillar/cmd/loguploader/loguploader.go#L580-L581) and modify [relative link](https://github.com/lf-edge/eve/pull/1667#discussion_r531123817)).

Comment (naiming-zededa) [reply pkg/server/apiHandler.go:316]: ok, let me change the url formatting to support the APIV1 correctly. Didn't realize adam was using that still.

**Comment by giggsoff** [comment]
@deitch please take a look at the second commit in this PR. The first one is yours from #51

File: pkg/driver/file/device_manager_file.go [hunk 4]
```diff
@@ -955,7 +955,7 @@ func (d *DeviceManager) getOnboardPath(cn string) string {
 
 func openTimestampFile(filename string) (*os.File, error) {
 	// open a new one
-	fullPath := path.Join(filename, time.Now().Format("2006-01-02T15:04:05"))
+	fullPath := path.Join(filename, time.Now().Format("2006-01-02T15:04:05.111"))
```
Comment (rvs) [inline pkg/driver/file/device_manager_file.go:958]: Why is this changed? @giggsoff

**Review by rvs — DISMISSED** [review]
LGTM

File: pkg/driver/file/device_manager_file.go (hunk 4 above)
Comment (giggsoff) [reply pkg/driver/file/device_manager_file.go:958]: We can receive events faster than one per second, so it looks reasonable to me.

Comment (rvs) [reply pkg/driver/file/device_manager_file.go:958]: Makes sense @giggsoff

**Review by deitch — APPROVED** [review]
LGTM. Thanks for pulling this together @giggsoff

**Comment by deitch** [comment]
Is this ready to go @giggsoff ?

**Comment by giggsoff** [comment]
Yes, it is ready.

**Comment by rvs** [comment]
Merged it. Do you need anything else @giggsoff to proceed with the Eden release?

**Comment by giggsoff** [comment]
I need a new docker image with it.

**Comment by rvs** [comment]
Seems like `lfedge/adam:0.0.66` is available -- can you please try it?
